batch/test/test_copy_files.py

```diff
@@ -0,0 +1,547 @@
+
+    def assert_batch_same_as_gsutil(self, src, dest):
+        result = _run_batch_same_as_gsutil(self.local_dir.name, self.remote_dir, src, dest)
+        assert all(_get_batch_gsutil_files_same(result)), str(result)
+        return result
+
+    def test_download_file_by_name(self):
+        self.assert_batch_same_as_gsutil('/data/a', '/')
+
+    def test_download_file_with_rename(self):
+        self.assert_batch_same_as_gsutil('/data/a', '/b')
+
+    def test_download_file_asterisk(self):
+        self.assert_batch_same_as_gsutil('/data/*', '/')
+
+    def test_download_file_match_brackets(self):
+        self.assert_batch_same_as_gsutil('/data/[ab]', '/b')
+
+    def test_download_file_question_mark(self):
+        self.assert_batch_same_as_gsutil('/data/?', '/')
+
+    def test_download_file_double_question_marks(self):
+        result = self.assert_batch_same_as_gsutil('/data/??', '/')
+        assert 'FileNotFoundError' in _get_output(result, 'rr', 'batch'), _get_output(result, 'rr', 'batch')
+        assert 'FileNotFoundError' in _get_output(result, 'rl', 'batch'), _get_output(result, 'rl', 'batch')
+        assert 'FileNotFoundError' in _get_output(result, 'lr', 'batch'), _get_output(result, 'lr', 'batch')
+        assert 'FileNotFoundError' in _get_output(result, 'll', 'batch'), _get_output(result, 'll', 'batch')
+
+    def test_download_multiple_files_to_single_file(self):
+        result = self.assert_batch_same_as_gsutil('/data/[ab]', '/b')
+        assert 'NotADirectoryError' in _get_output(result, 'rl', 'batch'), _get_output(result, 'rl', 'batch')
+        assert 'NotADirectoryError' in _get_output(result, 'll', 'batch'), _get_output(result, 'll', 'batch')
+
+    def test_download_file_invalid_dest_path_with_slash(self):
+        result = self.assert_batch_same_as_gsutil('/data/a', '/b/')
+        assert 'skipping destination file ending with slash' in _get_output(result, 'rl', 'batch'), _get_output(result, 'rl', 'batch')
+
+        # it's unclear why gsutil doesn't just create the directory like it does if the destination is remote
+        # it's also unclear why you don't get the same error as for the remote->local case
+        assert 'IsADirectoryError' in _get_output(result, 'll', 'batch'), _get_output(result, 'll', 'batch')
+
+    def test_download_file_invalid_dest_dir_with_wildcard(self):
+        result = self.assert_batch_same_as_gsutil('/data/*', '/b/')
+
+        assert 'destination must name a directory when matching multiple files' in _get_output(result, 'rl', 'batch'), _get_output(result, 'rl', 'batch')
+        assert 'NotADirectoryError' in _get_output(result, 'rl', 'batch'), _get_output(result, 'rl', 'batch')
+
+        assert 'destination must name a directory when matching multiple files' in _get_output(result, 'll', 'batch'), _get_output(result, 'll', 'batch')
+        assert 'NotADirectoryError' in _get_output(result, 'll', 'batch'), _get_output(result, 'll', 'batch')
```

Comment:
The last there methods don't assert anything about rr or lr, is that intentional?

Reply:
We check everything is the same in `self.assert_batch_same_as_gsutil`. The point of the additional asserts is to make sure I actually printed the error message I intended and not a different error message.